From 8b18cb6883ea66388a8bc26e9b6474be02e9dd91 Mon Sep 17 00:00:00 2001
From: dummy <dummy@dummy.dummy>
Date: Mon, 4 Jul 2016 13:01:40 -0700
Subject: [PATCH] v4v XSM

Extracted from current OpenXT code and put down on top of v9 v4v.

Signed-off-by: Christopher Clark <christopher.clark@baesystems.com>
---
 xen/include/xsm/dummy.h               | 5 +++++
 xen/include/xsm/xsm.h                 | 6 ++++++
 xen/xsm/dummy.c                       | 1 +
 xen/xsm/flask/hooks.c                 | 6 ++++++
 xen/xsm/flask/policy/access_vectors   | 5 +++++
 xen/xsm/flask/policy/security_classes | 1 +
 6 files changed, 24 insertions(+)

diff --git a/xen/include/xsm/dummy.h b/xen/include/xsm/dummy.h
index f65fa9a..7b4e7aa 100644
--- a/xen/include/xsm/dummy.h
+++ b/xen/include/xsm/dummy.h
@@ -636,6 +636,11 @@ static XSM_INLINE int xsm_ioport_mapping(XSM_DEFAULT_ARG struct domain *d, uint3
     return xsm_default_action(action, current->domain, d);
 }
 
+static XSM_INLINE int xsm_v4v_send(struct domain *d, struct domain *t)
+{
+    return 0;
+}
+
 #endif /* CONFIG_X86 */
 
 #ifdef CONFIG_ARM
diff --git a/xen/include/xsm/xsm.h b/xen/include/xsm/xsm.h
index 7ab80f8..76da5d7 100644
--- a/xen/include/xsm/xsm.h
+++ b/xen/include/xsm/xsm.h
@@ -165,6 +165,7 @@ struct xsm_operations {
 #ifdef CONFIG_ARM
     int (*map_gmfn_foreign) (struct domain *d, struct domain *t);
 #endif
+    int (*v4v_send) (struct domain *dom1, struct domain *dom2);
 };
 
 #ifdef XSM_ENABLE
@@ -624,6 +625,11 @@ static inline int xsm_ioport_mapping (xsm_default_t def, struct domain *d, uint3
 {
     return xsm_ops->ioport_mapping(d, s, e, allow);
 }
+
+static inline int xsm_v4v_send(struct domain *d1, struct domain *d2)
+{
+    return xsm_ops->v4v_send(d1,d2);
+}
 #endif /* CONFIG_X86 */
 
 #ifdef CONFIG_ARM
diff --git a/xen/xsm/dummy.c b/xen/xsm/dummy.c
index bc2e28e..0d0ac95 100644
--- a/xen/xsm/dummy.c
+++ b/xen/xsm/dummy.c
@@ -135,4 +135,5 @@ void xsm_fixup_ops (struct xsm_operations *ops)
     set_to_dummy_if_null(ops, map_gmfn_foreign);
 #endif
     set_to_dummy_if_null(ops, memory_translate);
+    set_to_dummy_if_null(ops, v4v_send);
 }
diff --git a/xen/xsm/flask/hooks.c b/xen/xsm/flask/hooks.c
index a30a782..04169e3 100644
--- a/xen/xsm/flask/hooks.c
+++ b/xen/xsm/flask/hooks.c
@@ -1470,6 +1470,11 @@ static int flask_map_gmfn_foreign(struct domain *d, struct domain *t)
 }
 #endif
 
+static int flask_v4v_send(struct domain *dom1, struct domain *dom2)
+{
+    return domain_has_perm(dom1, dom2, SECCLASS_V4V, V4V__SEND);
+}
+
 long do_flask_op(XEN_GUEST_HANDLE_PARAM(xsm_op_t) u_flask_op);
 
 static struct xsm_operations flask_ops = {
@@ -1578,6 +1583,7 @@ static struct xsm_operations flask_ops = {
 #ifdef CONFIG_ARM
     .map_gmfn_foreign = flask_map_gmfn_foreign,
 #endif
+    .v4v_send = flask_v4v_send,
 };
 
 static __init int flask_init(void)
diff --git a/xen/xsm/flask/policy/access_vectors b/xen/xsm/flask/policy/access_vectors
index 77b3b50..6329202 100644
--- a/xen/xsm/flask/policy/access_vectors
+++ b/xen/xsm/flask/policy/access_vectors
@@ -450,3 +450,8 @@ class security
 # remove ocontext label definitions for resources
     del_ocontext
 }
+
+class v4v
+{
+    send
+}
diff --git a/xen/xsm/flask/policy/security_classes b/xen/xsm/flask/policy/security_classes
index ef134a7..560f76e 100644
--- a/xen/xsm/flask/policy/security_classes
+++ b/xen/xsm/flask/policy/security_classes
@@ -17,5 +17,6 @@ class shadow
 class event
 class grant
 class security
+class v4v
 
 # FLASK
-- 
2.1.4

